### PR TITLE
bugfix/3185 transpose

### LINKED ISSUE #2903: Batchsize and learning rate scheduler
I was wondering if I need to adjust the batch size when using TPUs. I had a memory error when trying to run a (image 256x256x3) batch of size 128, which works perfectly fine on GPUs. 

Furthermore, do I need to adjust my custom learning rate scheduler which on GPUs run every batch (not just epoch):
```
def configure_optimizers(self):
        optimizer = get_optimizer(self)
        one_cycle_scheduler = OneCycleScheduler(optimizer, EPOCHS * steps_per_epoch // GRAD_ACCUMULATE)
        scheduler = {'scheduler': one_cycle_scheduler, "interval": "step"}

        return [optimizer], [scheduler]
```

pl version: 0.8.5
system: kaggle

### LINKED ISSUE #3185: Value out of range (expected to be in range of [-1, 0], but got 1)
## Value out of range (expected to be in range of [-1, 0], but got 1)
**Exception in device=TPU:5: torch_xla/csrc/helpers.cpp:97 : Check failed: min_shape_dim <= dim && dim <= max_shape_dim**

Following is the stack trace when I am using all 8 TPU cores on Kaggle. The exact same code with 

> sync_dist=False 

works completely fine on Kaggle GPU.

 

> Value out of range (expected to be in range of [-1, 0], but got 1)Exception in device=TPU:5: torch_xla/csrc/helpers.cpp:97 : Check failed: min_shape_dim <= dim && dim <= max_shape_dim 
> *** Begin stack trace ***
> 	tensorflow::CurrentStackTrace()
> 	torch_xla::XlaHelpers::GetCanonicalDimensionIndex(long long, long long)
> 	torch_xla::XlaHelpers::MakeTransposePermutation(long long, long long, long long)
> 	torch_xla::XLATensor::transpose(torch_xla::XLATensor const&, long long, long long)
> 	torch_xla::AtenXlaType::t(at::Tensor const&)
> 	c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<at::Tensor (*)(at::Tensor const&), at::Tensor, c10::guts::typelist::typelist<at::Tensor const&> >, at::Tensor (at::Tensor const&)>::call(c10::OperatorKernel*, at::Tensor const&)
> 	
> 	at::t(at::Tensor const&)
> 	
> 	
> 	
> 	at::Tensor::t() const
> 	
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyMethodDescr_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallDict
> 	_PyObject_Call_Prepend
> 	PyObject_Call
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallDict
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallDict
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallDict
> 	_PyObject_Call_Prepend
> 	
> 	_PyObject_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallDict
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	PyEval_EvalCodeEx
> 	PyEval_EvalCode
> 	
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyCFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyGen_Send
> 	_PyEval_EvalFrameDefault
> 	_PyGen_Send
> 	_PyEval_EvalFrameDefault
> 	_PyGen_Send
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyMethodDescr_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallDict
> 	_PyObject_Call_Prepend
> 	PyObject_Call
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyCFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyCFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyCFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyEval_EvalCodeWithName
> 	_PyFunction_FastCallDict
> 	_PyObject_Call_Prepend
> 	PyObject_Call
> 	_PyEval_EvalFrameDefault
> 	_PyGen_Send
> 	_PyMethodDef_RawFastCallKeywords
> 	_PyMethodDescr_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallKeywords
> 	_PyEval_EvalFrameDefault
> 	_PyFunction_FastCallDict
> *** End stack trace ***
> 

Here is the code for my LightningModule

```python
class GraphAEStage1(pl.LightningModule):
  def __init__(self, config):
    super(GraphAEStage1, self).__init__()
    max_deg = config['poc_max_degree']
    input_dim = config['input_dim']
    hidden_feat = config['hidden_feat_dim']
    layerid = config['layerid']
    self.lr = config['lr']
  
    self.encoder = GraphEncoderStage1(max_deg, input_dim, hidden_feat[layerid])
    self.decoder = GraphDecoderStage1(max_deg, input_dim, hidden_feat[layerid])
    
    self.encoder.apply(self.init_weights)
    
    self.crit = torch.nn.L1Loss()
    self.node_mae = MAE()
    self.adj_mae = MAE()
    
    self.decoder.self_linear.weight = nn.Parameter(self.encoder.self_linear.weight.permute(1,0))
    for d in range(config['poc_max_degree']):
        self.decoder.neigh_linear[d].weight = nn.Parameter(self.encoder.neigh_linear[d].weight.permute(1,0))

  def training_step(self, batch, batch_idx):
    loss, tf_node, _, _, _ = self.shared_step(batch)
    result = pl.TrainResult(minimize=loss)
    result.log('train_loss', loss, prog_bar = True, logger = True, on_step = False, on_epoch = True, sync_dist = True)
    return result

  def validation_step(self, batch, batch_idx):
    loss, tf_node, recon_node, val_node_mae, val_adj_mae = self.shared_step(batch)
    
    logs = {'val_loss': loss, 'node_mae': val_node_mae, 'adj_mae': val_adj_mae}
    result = pl.EvalResult(checkpoint_on = loss)
    result.log_dict(logs, prog_bar = True, logger = True, on_step = False, on_epoch = True, sync_dist = True)
    
    return result               

  def shared_step(self, batch):
    node, adj, deg, mask = batch
    tf_node, avg_adj = self.encoder(node, adj, deg, mask)
    recon_node, recon_adj = self.decoder(tf_node, deg)

    
    node_loss = self.criterion(recon_node, node)
    adj_loss = self.criterion(recon_adj, avg_adj)
    loss = node_loss + adj_loss
    
    node_mae = self.node_mae(node, recon_node)
    adj_mae = self.adj_mae(recon_adj, avg_adj)
    
    return loss, tf_node, recon_node, node_mae, adj_mae

  def configure_optimizers(self):
    optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
    return optimizer

  def criterion(self, output, target):
    return self.crit(output, target)

  def init_weights(self, m):
    if type(m) == nn.Linear:
        torch.nn.init.xavier_uniform(m.weight)
        m.bias.data.fill_(0.00)  
```

There are similar issues reported on PyTorch XLA repo, but many of them were using CrossEntropyLoss. I am not using CrossEntropyLoss here.

### PR DESCRIPTION
**Pull request by lezwon**
## What does this PR do?

Fixes #3185
Fixes #2903 

`tensor.t()` breaks on XLA for 1-D tensors. Replaced it with `tensor.transpose(-1, 0)`.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by williamFalcon** [comment]
@lezwon mind adding a test so this doesn’t happen again?

The test is to use some of the step result tests with TPUs which you can copy over into the TPU section

**Comment by lezwon** [comment]
@williamFalcon I had to make a couple of tweaks to get the tests working. They aren't related to this issue but might have been bugs needed to be fixed.

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -229,7 +229,7 @@ def _stop_distributed_training(self, trainer, pl_module):
 
         if trainer.use_tpu:
             stop = torch.tensor(int(trainer.should_stop), device=pl_module.device, dtype=torch.int32)
-            stop = xm.mesh_reduce("stop_signal", stop, torch.cat)
+            stop = xm.mesh_reduce("stop_signal", stop, sum)
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/early_stopping.py:232]: does this solve https://github.com/PyTorchLightning/pytorch-lightning/issues/2903 issue??

**Review by Borda — COMMENTED** [review]
LGTM 🐰

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (lezwon) [reply pytorch_lightning/callbacks/early_stopping.py:232]: I don't think so. This issue was related to the `stop` tensor being a 0-D tensor. Running `torch.cat` on a list of 0-D resulted in an error `RuntimeError: zero-dimensional tensor (at position 0) cannot be concatenated`. I think `sum` was intended here instead.

Comment (lezwon) [reply pytorch_lightning/callbacks/early_stopping.py:232]: oh, wait. It does fix this issue on the notebook mentioned there.

Screenshot 2020-08-30 at 9 41 59 AM

Comment (rohitgr7) [reply pytorch_lightning/callbacks/early_stopping.py:232]: There is a new error coming now in the notebook, but I guess that is a different issue.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/callbacks/early_stopping.py:232]: wondering how is it possible that the error comes in Notebook but here it passes

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/early_stopping.py:232]: not sure, but I have observed tpu training usually fails either during the last epoch(>1) or 2nd epoch. but in tests `max_epochs` is set to 1.

**Pull request merged**